### PR TITLE
auth: De-dupe token refreshes into a single task

### PR DESCRIPTION
**Pull request by ParkMyCar**
This PR updates our Frontegg Auth client to take ownership of refreshing tokens, and spawning a single task to manage the refresh for a group of de-duplicated requests.

In #21801 we started de-duplicating authentication requests. This had the unintended consequence of breaking refresh tokens in some cases. A refresh token is returned from Frontegg upon successful auth and used to extend the expiration of a JWT. The refresh tokens are single use though. When we started de-duplicating requests we ended up with multiple sessions having the same refresh token, and all of them trying to use it. What ended up happening is a single session was able to refresh and all of the others were disconnected.

This PR updates the logic in our Frontegg client to continue using the same de-duplication strategy (i.e. an `Arc<Mutex<HashMap<...>>>`) but now we spawn a task to automatically refresh tokens and give the caller a channel which we push updates to.

It also adds a few metrics around Frontegg Auth.

I tested this PR manually, but still TODO is add tests, but I wanted to get this PR up for early feedback if possible. Also note, my first attempt at solving this made everything task based, I got rid of the `Arc<Mutex<HashMap<...>>>`. I think this is a cleaner approach but I wasn't able to get it working with our existing integration tests with are half synchronous and half asynchronous.

### Motivation

Fixes: https://github.com/MaterializeInc/materialize/issues/22096

### Tips for reviewer

This PR is split into three commits which might make it easier to review:

1. Refactor, moving `ExternalUserMetadata` into `mz_repr`.
2. Updates to the `mz_frontegg_auth` crate itself, this is the **most important commit**!
3. Update tests and callers of Authentication types to use the new API
4. Add tests to exercise duplicate sessions being refreshed together

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - Fixes a bug where sessions using the same app password that start immediately after one another, could get invalidated early.

### CODE REVIEW

File: src/adapter/src/session.rs [hunk 1]
```diff
@@ -694,18 +693,32 @@ impl<T: TimestampManipulation> Session<T> {
...
+            return;
+        };
+
+        // If the value hasn't changed then return.
+        if !rx.has_changed().unwrap_or(false) {
+            return;
         }
+
+        // Update our metadata!
+        let metadata = rx.borrow_and_update().clone();
```
Comment (guswynn) [inline src/adapter/src/session.rs:720]:
```suggestion
        // Update our metadata! Note the short critical section (just a clone), to avoid blocking the sender
        let metadata = rx.borrow_and_update().clone();
```

File: src/frontegg-auth/src/auth.rs [hunk 2]
```diff
@@ -7,82 +7,269 @@
...
+                    rx
+                }
+                // New request! Need to queue one up.
+                None => {
+                    tracing::debug!(?req, "spawning new request");
+
+                    inflight.insert(req.clone(), vec![tx]);
+
+                    let client = self.client.clone();
+                    let inflight = Arc::clone(&self.inflight_auth_requests);
```
Comment (guswynn) [inline src/frontegg-auth/src/auth.rs:187]: i think this implicitly drops the previous lock guard...can you make that explicit?

File: src/frontegg-auth/src/auth.rs [hunk 3]
```diff
@@ -117,115 +304,225 @@ impl Authentication {
...
+    expected_email: String,
+    validation_config: ValidationConfig,
+    refresh_tx: watch::Sender<ExternalUserMetadata>,
+) -> RefreshTaskId {
+    let refresh_url = format!("{}{}", admin_api_token_url, REFRESH_SUFFIX);
+    let id = RefreshTaskId::new();
+
+    // Continuously refresh.
+    let name = format!("frontegg-auth-refresh-{}", id);
+    mz_ore::task::spawn(|| name, async move {
```
Comment (guswynn) [inline src/frontegg-auth/src/auth.rs:408]: the nested spawn is a bit confusing but nothing seems wrong with it

File: src/frontegg-auth/src/auth.rs [hunk 4]
```diff
@@ -117,115 +304,225 @@ impl Authentication {
...
-                };
+/// Returns a [`Duration`] for how long the provided claims are valid for.
+fn valid_for(now: &NowFn, claims: &ValidatedClaims) -> Duration {
+    let valid_for = claims.exp - now.as_secs();
+    let valid_for = u64::try_from(valid_for).unwrap_or(0);
+
+    Duration::from_secs(valid_for)
+}
+
+/// Returns a `Future` that resolves then the sending side of the provided channel closes.
```
Comment (mjibson) [inline src/frontegg-auth/src/auth.rs:383]: `then` -> `when`

File: src/adapter/src/session.rs [hunk 5]
```diff
@@ -695,18 +693,32 @@ impl<T: TimestampManipulation> Session<T> {
...
+    /// Register a receiver which pushes updates of [`ExternalUserMetadata`].
+    pub fn register_external_metadata_transmitter(
         &mut self,
-    ) -> UnboundedSender<ExternalUserMetadata> {
-        self.external_metadata_tx.clone()
+        rx: tokio::sync::watch::Receiver<ExternalUserMetadata>,
+    ) {
+        let old = self.external_metadata_rx.replace(rx);
+        if old.is_some() {
+            tracing::error!("Double register of external metadata transmitter!");
```
Comment (mjibson) [inline src/adapter/src/session.rs:703]: What are the implications if this happens? Should we just panic because it means something has gone quite wrong?

File: src/frontegg-auth/src/auth.rs [hunk 6]
```diff
@@ -117,115 +304,222 @@ impl Authentication {
...
-                                //
-                                // TODO: report Frontegg errors at `error!`
-                                // level or via a Prometheus metric so that
-                                // we can alert on them.
-                                info!(
-                                    "failed to refresh token for {}: {:#}",
-                                    response.claims.email, e
-                                );
-                                // Retry again later. 5 seconds chosen arbitrarily.
-                                time::sleep(Duration::from_secs(5)).await;
```
Comment (mjibson) [inline src/frontegg-auth/src/auth.rs:210]: Where did this retry go? Is it still needed?

File: src/adapter/src/session.rs (hunk 1 above)
Comment (ParkMyCar) [reply src/adapter/src/session.rs]: Updated!

File: src/frontegg-auth/src/auth.rs (hunk 2 above)
Comment (ParkMyCar) [reply src/frontegg-auth/src/auth.rs:189]: Good call, I made this an explicit drop

File: src/frontegg-auth/src/auth.rs (hunk 3 above)
Comment (ParkMyCar) [reply src/frontegg-auth/src/auth.rs:410]: Yeah I'm not the biggest fan of it, but I don't think there is a way to avoid it in the current setup

File: src/frontegg-auth/src/metrics.rs [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+    pub refresh_tasks_active: IntGaugeVec,
+}
+
+impl Metrics {
+    pub(crate) fn register_into(registry: &MetricsRegistry) -> Self {
+        Self {
+            request_count: registry.register(metric!(
+                name: "mz_auth_request_count",
+                help: "Total number of HTTP requests made to Frontegg for authentication",
+                var_labels: ["path", "status"],
```
[the remark this answers is not in the document]
Comment (ParkMyCar) [reply src/frontegg-auth/src/metrics.rs:33]: Because we technically hit two different endpoints, token exchange and refresh. I figured differentiating between the two would be good, I can remove this of you want!

File: src/frontegg-auth/src/auth.rs (hunk 4 above)
Comment (ParkMyCar) [reply src/frontegg-auth/src/auth.rs]: Done!

File: src/adapter/src/session.rs (hunk 5 above)
Comment (ParkMyCar) [reply src/adapter/src/session.rs]: Good question! the implications are the session not getting updated external metadata. Which means a user could lose admin privileges but the session would not get updated. I turned this into an error which forces the caller to handle the scenario. We could panic but I wanted to be a bit defensive, what do you think?

File: src/frontegg-auth/src/auth.rs (hunk 6 above)
Comment (ParkMyCar) [reply src/frontegg-auth/src/auth.rs:210]: All retry-able errors will already get retried by the underlying client, so this logic was no longer needed.

File: src/frontegg-auth/src/metrics.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (mjibson) [reply src/frontegg-auth/src/metrics.rs:33]: Ah, I didn't read carefully enough and thought these were our HTTP endpoints. Ok.

**Review by mjibson — APPROVED** [review]

File: src/adapter/src/session.rs (hunk 5 above)
Comment (mjibson) [reply src/adapter/src/session.rs]: lgtm
